BlockIcon, DropDown panes, and other fun things

uecasm opened the pull request:
This is an omnibus PR best reviewed commit-by-commit.  I could split it to separate PRs if you really want but I'm hopeful it's sufficiently uncontroversial as-is.

Bugfixes:
* Fixes incorrect spacing of child elements in lists when `childspacing` is set (typically to -1 to avoid double-thickness lines).
* Fixes tooltips being stuck permanently visible when the parent of their owner is hidden (but not their owner directly).

Features:
* Adds a `blockicon` control that is similar to `itemicon` and `entityicon` but displays `BlockState` directly.
    ![image](https://github.com/ldtteam/BlockUI/assets/539951/d2e08960-4db4-481d-9ca8-a7a12b62b9e4)
* The `dropdown` now supports the base `DataProvider` in addition to the string-only one -- this allows more complex child content than simply a list of text-labeled buttons.  It should still be compatible with other existing uses, though I only tested it with Structurize and not yet MineColonies.
    ![image](https://github.com/ldtteam/BlockUI/assets/539951/f20d83c8-d715-44a9-ba08-91c484ace576)
* When a `button` is specified with `source=""` this now displays as an invisible button instead of throwing.  This is useful to allow layering a button on top of other controls to make things clickable that otherwise wouldn't be (and is key to the new dropdown support).
* Adds the `horizontal="true"` option to a `group`, which does what it sounds like.
* Any pane can be marked as `cloneable="true"`, which allows the `clone(View)` method to be called to create a copy of (the original state of) that control, similar to what the `list` control does.

This is currently 1.20-only and should not be backported as is.  I have no current plans to backport it but I could be persuaded.

uecasm committed:
BlockIcon

uecasm committed:
Allow ButtonImage to be invisible

uecasm committed:
DropDownList supports generic views

uecasm committed:
Update test window

uecasm committed:
Fix tooltips getting stuck when parent is hidden

uecasm committed:
Fix incorrect list spacing

uecasm commented:
There is a minor rendering bug with this, which I'm not sure if it's BlockIcon-specific or BlockUI in general: if something else tries to render above the icons (e.g. a dropdown menu from another control) then there are some small "cutouts" where it draws some of the content above but not all of it (e.g. it will draw background but not text, or draw some of another icon but with some holes).  e.g. (note artifacts on the `t` and the last `e`):
![image](https://github.com/ldtteam/BlockUI/assets/539951/b2df93d6-a4d1-4b9c-83be-dccb19a7448f)

This is sufficiently minor that I think it's mostly ok as is, but if anyone has ideas on how to resolve it, I'm all ears.

Nightenom commented on src/main/java/com/ldtteam/blockui/views/ScrollingListContainer.java:
```diff
@@ -30,7 +30,7 @@ public void refreshElementPanes(final ScrollingList.DataProvider dataProvider, f
                 final Pane child;
-                final int childYpos = (childSpacing + listElementHeight) * i;
+                final int childYpos = listElementHeight * i;
                 if (childYpos + listElementHeight >= scrollY && childYpos <= scrollY + height)
```
```suggestion
                if (childYpos + (listElementHeight - childSpacing) >= scrollY && childYpos <= scrollY + height)
```
since you are here (from discord discussion)

Nightenom reviewed (changes requested):
Remove blockicon, will pr myself with better way using vanilla systems
Are you 100% sure with backward stability of dropdown changes?

uecasm committed:
Fix per review

uecasm committed:
Fix some lighting and rotation derps

uecasm commented:
Fixed some errors in the rotation for BlockIcon.  (I'd rather keep it in until there's an alternative.)  This also fixes most of the issue with an overlaid menu, but not entirely.  (The text menu is fixed, but a dropdown of BlockIcons on top of other BlockIcons still misbehaves.)

Regarding the dropdown, afaik the only incompatibility is that if you declare a dropdown element and *don't* put an `id="button"` inside, then it used to be able to display a current value on the base button (but couldn't display anything in the dropdown itself).  Now it will more consistently not display anything in either location.  At least in Structurize and MineColonies, there are no such usages, because that would be fairly useless anyway.  I have not yet tested this with MineColonies, as stated, but I have tested it with an existing dropdown in Structurize and it worked fine.

uecasm committed:
Horizontal Group

uecasm committed:
Pick #52 except for ItemIcon

Update BlockIcon to use new renderer (sacrificing rotation 😢)

uecasm commented:
I've merged most of #52 into here (excluding the changes to `ItemIcon`, since they make no sense to me) and changed `BlockIcon` to use your rendering code instead -- sadly losing the ability to rotate the block (though that wasn't needed for my use case, it was nice to have).

Also added a couple of other useful things needed downstream.  Technically the `dropdown` changes are no longer required, so I could drop them if you don't like them, but they still seem useful and safe so I'm inclined to leave it.

Nightenom commented:
Closing this:
- blockicon, hover and other small fixes were PRed elsewhere
- cloning should be done through moving loading code into codecs (kinda hard work - especially xml ops, but should be eventually done)
- horizonal group and the dropdown thing can be PRed again if wanted (please separate them)
